collect/collect_test.go

```diff
@@ -2679,75 +2661,105 @@ func BenchmarkCollectorWithSamplers(b *testing.B) {
+
+		data["service.name"] = fmt.Sprintf("service-%d", spanIdx%3)
+		data["duration_ms"] = 100.0
+
+		data["meta.signal_type"] = "trace"
+		data["meta.annotation_type"] = "span"
+		data["meta.refinery.incoming_user_agent"] = "refinery/v2.4.1"
+
+		// Large text fields (3 large fields)
+		data["large_field_1"] = largeText
+		data["large_field_2"] = mediumText
+		data["large_field_3"] = smallText
+
+		// many string, float, int, and bool fields
+		for j := 0; j < 25; j++ {
+			data[fmt.Sprintf("string.field_%d", j)] = fmt.Sprintf("string_value_%d", j)
+			data[fmt.Sprintf("int.field_%d", j)] = int64(j)
+			data[fmt.Sprintf("float.field_%d", j)] = float64(j)
+			data[fmt.Sprintf("bool.field_%d", j)] = (spanIdx+j)%2 == 0
+		}
 
-	// fields used for sampling
-	data["sampler-field-1"] = rand.Intn(20)
-	data["sampler-field-2"] = "static-value"
+		// Array field
+		data["tags"] = []string{
+			fmt.Sprintf("tag_%d", spanIdx),
+			fmt.Sprintf("tag_%d", spanIdx+1),
+			fmt.Sprintf("tag_%d", spanIdx+2),
+		}
 
-	data["trace.trace_id"] = traceID
-	data["trace.span_id"] = spanID
-	data["trace.parent_id"] = parentID
-	data["service.name"] = fmt.Sprintf("service-%d", spanIdx%3)
-	data["duration_ms"] = 100.0
+		// Nested map field
+		data["custom.metadata"] = map[string]interface{}{
+			"nested_field_1": map[string]interface{}{
+				"key1": fmt.Sprintf("value_%d", spanIdx),
+				"key2": spanIdx * 10,
+				"key3": spanIdx%2 == 0,
+			},
+			"nested_field_2": map[string]interface{}{
+				"key1": true,
+				"key2": false,
+				"key3": spanIdx + 100,
+			},
+		}
+		payload := types.NewPayload(cfg, data)
+		payload.ExtractMetadata()
 
-	data["meta.signal_type"] = "trace"
-	data["meta.annotation_type"] = "span"
-	data["meta.refinery.incoming_user_agent"] = "refinery/v2.4.1"
+		span = &types.Span{
+			Event: types.Event{
+				Dataset: "benchmark-dataset",
+				APIKey:  "test-api-key",
+				Data:    payload,
+			},
+		}
+	}
 
-	// Large text fields (3 large fields)
-	data["large_field_1"] = largeText
-	data["large_field_2"] = mediumText
-	data["large_field_3"] = smallText
+	span.TraceID = traceID
+	span.IsRoot = isRoot
 
-	// many string, float, int, and bool fields
-	for j := 0; j < 25; j++ {
-		data[fmt.Sprintf("string.field_%d", j)] = fmt.Sprintf("string_value_%d", j)
-		data[fmt.Sprintf("int.field_%d", j)] = int64(j)
-		data[fmt.Sprintf("float.field_%d", j)] = float64(j)
-		data[fmt.Sprintf("bool.field_%d", j)] = (spanIdx+j)%2 == 0
+	// Core tracing fields (always present, differs per trace)
+	parentID := ""
+	if spanIdx%3 != 0 { // 2/3 of spans have parents
+		parentID = fmt.Sprintf("span-%018d", spanIdx-1)
 	}
-
-	// Array field
-	data["tags"] = []string{
-		fmt.Sprintf("tag_%d", spanIdx),
-		fmt.Sprintf("tag_%d", spanIdx+1),
-		fmt.Sprintf("tag_%d", spanIdx+2),
+	span.Event.Data.Set("trace.trace_id", traceID)
+	span.Event.Data.Set("trace.span_id", fmt.Sprintf("span-%018d", spanIdx))
+	if isRoot {
+		span.Event.Data.UnsetForTest("trace.parent_id")
```

Comment:
If it's root, I think we also need to unset the `IsRoot` field on the span as well as the `MetaIsRoot` field on the `Payload`